stern/stern.go

```diff
@@ -185,50 +161,41 @@ func Run(ctx context.Context, config *Config) error {
 						return
 					}
 					added <- v
-				case v, ok := <-r:
-					if !ok {
-						errCh <- fmt.Errorf("lost watch connection")
-						return
-					}
-					removed <- v
 				case <-ctx.Done():
 					return
 				}
 			}
 		}()
 	}
 
-	go func() {
-		for p := range added {
-			targetID := p.GetID()
-
-			if tail, ok := getTail(targetID); ok {
-				if tail.isActive() {
-					continue
-				} else {
-					tail.Close()
-					clearTail(targetID)
-				}
+	addTarget := func(ctx context.Context, target *Target) {
+		for {
+			start := time.Now()
+			tail := newTail(target)
+			err := tail.Start(ctx)
+			tail.Close()
+			if err == nil {
+				return
 			}
-
-			tail := newTail(p)
-			setTail(targetID, tail)
-
-			go func(tail *Tail) {
-				if err := tail.Start(ctx); err != nil {
-					fmt.Fprintf(config.ErrOut, "unexpected error: %v\n", err)
+			if !filter.isActive(target) {
+				fmt.Fprintf(config.ErrOut, "failed to tail: %v\n", err)
+				return
+			}
+			if interval := time.Until(start.Add(1 * time.Minute)); interval > 0 {
```

Comment:
What should we do with the retry interval? I don't think the error occurs frequently, so I think it's a problem that doesn't need to be considered deeply. Can we consider an exponential backoff of up to 1 minute?

Reply:
Thank you for pointing it out.

There would be two cases when we need to retry.

1. It fails when it starts
    - e.g., lack of the `pods/log` permission
2. It is disconnected on the way
    - Load balancers or firewalls might disconnect when there are no logs for a while

We need to prevent a burst of retries in case 1. On the other hand, I think it is better to retry immediately in case 2. (Case 2 often happens in my use case :cry:.)

What do you think about using [a rate limiter](https://pkg.go.dev/golang.org/x/time/rate) to support the above two cases? We can consider that if it takes 10 seconds or more, it is case 2.

https://go.dev/play/p/tdNp9huQmSI

```go
	addTarget := func(ctx context.Context, target *Target) {
		limiter := rate.NewLimiter(rate.Every(time.Second*10), 3)
		for {
			if err := limiter.Wait(ctx); err != nil {
				fmt.Fprintf(config.ErrOut, "failed to retry: %v\n", err)
				return
			}
			tail := newTail(target)
			err := tail.Start(ctx)
			tail.Close()
			if err == nil {
				return
			}
			if !filter.isActive(target) {
				fmt.Fprintf(config.ErrOut, "failed to tail: %v\n", err)
				return
			}
			fmt.Fprintf(config.ErrOut, "failed to tail: %v, will retry\n", err)
		}
	}
```

Reply:
rate limiter looks good!

Reply:
Thank you! I changed it in https://github.com/stern/stern/pull/221/commits/01dedcad56fbb135043e7ae8c71fba98ca38d1d8 .